crates/bevy_tasks/src/thread_executor.rs

```diff
@@ -77,33 +77,48 @@ impl<'task> ThreadExecutor<'task> {
     pub fn ticker<'ticker>(&'ticker self) -> Option<ThreadExecutorTicker<'task, 'ticker>> {
         if thread::current().id() == self.thread_id {
             return Some(ThreadExecutorTicker {
-                executor: &self.executor,
+                executor: self,
                 _marker: PhantomData::default(),
             });
         }
         None
     }
+
+    /// check whether `self` and `other` is the same executor
+    fn is_same_executor(&self, other: &Self) -> bool {
+        std::ptr::eq(self, other)
+    }
 }
 
 /// Used to tick the [`ThreadExecutor`]. The executor does not
 /// make progress unless it is manually ticked on the thread it was
 /// created on.
 #[derive(Debug)]
 pub struct ThreadExecutorTicker<'task, 'ticker> {
-    executor: &'ticker Executor<'task>,
+    executor: &'ticker ThreadExecutor<'task>,
     // make type not send or sync
     _marker: PhantomData<*const ()>,
 }
 impl<'task, 'ticker> ThreadExecutorTicker<'task, 'ticker> {
     /// Tick the thread executor.
     pub async fn tick(&self) {
-        self.executor.tick().await;
+        self.executor.executor.tick().await;
     }
 
     /// Synchronously try to tick a task on the executor.
     /// Returns false if if does not find a task to tick.
     pub fn try_tick(&self) -> bool {
-        self.executor.try_tick()
+        self.executor.executor.try_tick()
+    }
+
+    /// join tick `self` and `other` with `or`.
+    /// check [this pr](https://github.com/bevyengine/bevy/pull/7825) for reference
+    pub async fn or_tick(&self, other: &Self) {
+        if self.executor.is_same_executor(other.executor) {
```

Comment:
In my other comment I meant that you should move this if into the TaskPool::scope function. There is already a bunch of branching being done there with picking which executors to tick/run. Putting this branch here is putting it back in the hot loop

Reply:
Sure, I think it is better to put all branching logic to one place. Though perf wise, I think the hot loop actually is not that hot, each iteration, it calls async_executor's code, which involved several atomic instructions.